10 Pearls with Azure Functions in Pulumi (#1557)
pulumi/docs: content/blog/10-pearls-with-azure-functions-in-pulumi/index.md

---

> title: "10 Pearls With Azure Functions in Pulumi"

Reviewer: ```suggestion
title: "Ten Pearls With Azure Functions in Pulumi"
```

date: 2019-08-15

> draft: true

Reviewer: change before publish?

Author: I still need a feature image... and the actual publish date

Author: Added a feature image

meta_image: meta.png
authors: ["mikhail-shilkov"]
tags: ["Azure", "Serverless"]
meta_desc: "In this post, we'll take a look at 10 pearls&mdash;bite-sized code snippets&mdash;that demonstrate using Pulumi to build serverless applications with Azure Functions and infrastructure as code."
---

In this post, we'll take a look at 10 "pearls"&mdash;bite-sized code snippets&mdash;that demonstrate using Pulumi to build serverless applications with Azure Functions and infrastructure as code. These pearls are organized into four categories, each demonstrating a unique scenario:

> - **Function App Deployment**: Take an existing Azure Functions application in your language of choice and deploy it with infrastructure as code.

Reviewer: ```suggestion
- **Function App Deployment**: Deploy an Azure Functions application using infrastructure as code.
```

Author: Don't you think this becomes too broad - the other three bullet points fit this shorter definition too.

> - **HTTP Functions as Callbacks**: Mix TypeScript or JavaScript functions right into the infrastructure definition to produce concise, strongly-typed, self-contained, serverless HTTP endpoints.

Reviewer: ```suggestion
- **HTTP Functions as Callbacks**: Mix JavaScript or TypeScript functions with your infrastructure definition to produce strongly-typed, self-contained, serverless HTTP endpoints.
```

> - **Cloud Event Handling**: Leverage the whole variety of event sources available to Azure Functions with lightweight and intuitive event handlers.

Reviewer: ```suggestion
- **Cloud Event Handling**: Leverage a variety of event sources available to Azure Functions with lightweight event handlers.
```

> - **Data Flows with Function Bindings**: Pull extra data into the functions and push results out with function bindings&mdash;declarative connectors to Azure services.

Reviewer: ```suggestion
- **Data Flows with Function Bindings**: Take advantage of function bindings&mdash;declarative connectors to Azure services.
```

<!--more-->

> Here is a complete index of the pearls including associated steps, in case you want to jump to a specific one:

Reviewer: ```suggestion
Our Azure Functions pearls are:
```

[**Function App Deployment**](#function-app-deployment)

- [Deploy a Function App written in .NET or any other supported runtime](#1-deploy-a-net-function-app)
- [Configure Functions to run on an Elastic Premium Plan](#2-run-functions-using-an-elastic-premium-plan)

[**HTTP Functions as Callbacks**](#http-functions-as-callbacks)

- [Define Node.js Functions as inline callbacks](#3-define-node-js-functions-as-inline-callbacks)
- [Implement REST APIs as multiple Functions](#4-rest-apis-as-multiple-functions)
- ["Warm" the Functions to avoid Cold Starts](#5-function-warming-with-a-timer-function)

[**Cloud Event Handling**](#cloud-event-handling)

- [Process events from Azure Event Hub](#6-process-events-from-azure-event-hubs)
- [Subscribe to Blob creation with Azure Event Grid](#7-subscribe-to-azure-event-grid)
- [Run a Function every time an Azure resource is modified](#8-respond-to-resource-level-events)

[**Data Flows with Function Bindings**](#data-flows-with-function-bindings)

- [Push a message to a Storage Queue with an output binding](#9-output-bindings)
- [Pull a Storage Table row with an input binding](#10-input-bindings)

## Function App Deployment

> Azure Functions support many languages, and Pulumi doesn't limit your choice. You are free to take any existing serverless application and deploy it in an infrastructure-as-code manner.

Reviewer: ```suggestion
Azure Functions support many languages, and Pulumi supports whatever choice you make. You can take any existing serverless application and deploy it using infrastructure-as-code.
```

Author: Two "support" in one sentence. "Pulumi respects"? "plays along with"?

### 1. Deploy a .NET Function App

[ [Runnable Example](https://github.com/pulumi/examples/tree/master/azure-ts-functions-raw) ]

> Many Function Apps are .NET applications created with native tooling like Visual Studio or the Functions CLI. With no required changes to the code, Pulumi can deploy such an application in several lines of JavaScript:

Reviewer: ```suggestion
Many Function Apps are .NET applications created with native tooling like Visual Studio or the Functions CLI. Pulumi can simply deploy such an application in only a few lines of JavaScript:
```

``` ts
const dotnetApp = new azure.appservice.ArchiveFunctionApp("http-dotnet", {
   resourceGroupName: resourceGroup.name,
   archive: new pulumi.asset.FileArchive("./app/bin/Debug/netcoreapp2.1/publish"),
   appSettings: { runtime: "dotnet" },
});
```

There are only four things required: 

- the Function App name ("http-dotnet")
- the resource group it belongs to
- the path to the compiled .NET assemblies, and

> - the desired runtime. 

Reviewer: ```suggestion
- the desired runtime
```

> Pulumi takes care of the rest for you.It handles the following tasks:

Reviewer: ```suggestion
Pulumi takes care of the rest for you. It handles the following tasks:
```

- Creating a Storage Account and a Blob Container
- Zipping up the binaries and uploading them to the blob container
- Calculating a SAS token

> - Preparing a Consumption Plan and a Function App on said Consumption Plan

Reviewer: ```suggestion
- Preparing a Consumption Plan and a Function App using this Consumption Plan
```

- Configuring the required application settings, including a reference to the zip archive with the SAS token

![Serverless Application deployed by Pulumi](console.png)

While a few values are required, you are not restricted to the default behavior and can change any setting.

### 2. Run Functions using an Elastic Premium Plan

> While the Consumption Plan is the ultimate serverless option, there are other ways to host Azure Functions. A while ago, Microsoft introduced the Premium Plan: a combination of the power and guarantees of a fixed App Service Plan with the elasticity of Consumption.

Reviewer: ```suggestion
While the Consumption Plan is the most popular serverless option, there are other ways to host Azure Functions. Azure also offers the Premium Plan: a combination of the power and guarantees of a fixed App Service Plan with the elasticity of Consumption.
```

   plan: premiumPlan,
   archive: new pulumi.asset.FileArchive("./java/target/functions/fabrikam"),
   appSettings: { runtime: "java" },
});
```

In this example, I deployed a Java application. The Premium Plan has a couple of configuration knobs: the *instance size* and the *maximum scale-out limit*.

## HTTP Functions as Callbacks

Node.js is another runtime supported by Azure Functions. Like before, you can use the `ArchiveFunctionApp` class to deploy the application from an external folder. However, Pulumi's Node.js SDK provides a way to mix the code of your functions directly into your infrastructure definition.

### 3. Define Node.js Functions as Inline Callbacks

[ [Runnable Example](https://github.com/pulumi/examples/tree/master/azure-ts-functions) ]

A TypeScript or a JavaScript function becomes an Azure Function deployed to the cloud:

``` ts
const greeting = new azure.appservice.HttpEventSubscription('greeting', {
   resourceGroupName: resourceGroup.name,
   callback: async (context, req) => {
       return {
           status: 200,
           body: `Hello ${req.query['name'] || 'World'}!`,
       };
   }
});

export const url = greeting.url;
```

> You are free to use any NPM module inside the callback. The dependencies are automatically packaged into the deployment artifact. Likewise, you can extract the callback function out into a separate file or package and import it from the infrastructure code.

Reviewer: ```suggestion
You are free to use any NPM module inside the callback and the dependencies will be automatically packaged into the deployment artifact. Alternatively, you can extract the callback function into a separate file or package and import it from your infrastructure code.
```

``` ts
const get = new azure.appservice.HttpFunction("Read", {
   route: "items",
   methods: ["GET"],
   callback: async (context, request) => {
       const items = await repository.list();
       return { status: 200, body: items };
   },
});

const post = new azure.appservice.HttpFunction("Add", {
   route: "items",
   methods: ["POST"],
   callback: async (context, request) => {
       const id = await repository.add(request.body);
       return { status: 201, body: { id }  };
   },
});

const app = new azure.appservice.MultiCallbackFunctionApp("multi-app", {
   resourceGroupName: resourceGroup.name,
   functions: [get, post],
});
```

You can combine as many Functions as you want, including functions of different types, as shown in the next example.

### 5. Function "Warming" with a Timer Function

[ [Runnable Example](https://github.com/pulumi/pulumi-azure/blob/master/examples/http-multi) ]

Scheduled jobs are another frequent use case for serverless functions. It's possible to define a [*cron expression*](https://docs.microsoft.com/azure/azure-functions/functions-bindings-timer#cron-expressions) and get the code executed at designated intervals.

> The Consumption Plan disposes a worker if no Function runs in about 20 minutes. The next execution causes a [*cold start*](https://mikhail.io/serverless/coldstarts/define/), and the response latency to be high.

Reviewer: ```suggestion
The Consumption Plan disposes a worker if no Function runs in about 20 minutes. After disposal, the next execution will cause a [*cold start*](https://mikhail.io/serverless/coldstarts/define/), and the response latency, will increase.
```

> The following example might seem unexpected, but it's a well-known pattern in the world of Azure Functions. We combine a target HTTP Function with a Timer Function executed once in several minutes. The body of the Timer Function is empty. Its sole purpose is to trigger the Function App at a regular cadence to keep the worker "warm":

Reviewer: ```suggestion
In order to avoid this disposal, we can combine a target HTTP Function with a Timer Function. The body of the Timer Function is empty, since its sole purpose is to trigger the Function App and keep the worker "warm":
```

``` ts
const warmer = new azure.appservice.TimerFunction("warmer", {
   schedule: "0 */5 * * * *",
   callback: async () => {},
});

const http = new azure.appservice.HttpFunction("hello", {
   callback: async (context, req) => {
       return {
           status: 200,
           body: "Hello World!",
       };
   }
});

const app = new azure.appservice.MultiCallbackFunctionApp("always-warm-app", {
   resourceGroupName: resourceGroup.name,
   functions: [http, warmer],
});
```

It's easy to imagine a custom component `WarmedFunctionApp` which appends a standard Timer Function to an array of Functions passed to its constructor.

## Cloud Event Handling

> While HTTP is a widespread use case, Azure Functions support many other trigger types too. [The previous post]({{< relref "serverless-as-simple-callbacks-with-pulumi-and-azure-functions" >}}) featured Storage Queues and ServiceBus Topics. Now, Pulumi supports Timers, Events Hubs, Event Grid, Storage Blobs, Service Bus Queues, and Cosmos DB Change Feed events, too! Let's see how that looks using Azure Event Hubs.

Reviewer: ```suggestion
While HTTP is a widespread use case, Azure Functions support many other trigger types too. [The previous post]({{< relref "serverless-as-simple-callbacks-with-pulumi-and-azure-functions" >}}) featured Storage Queues and ServiceBus Topics. Pulumi supports Timers, Events Hubs, Event Grid, Storage Blobs, Service Bus Queues, and Cosmos DB Change Feed events, too! Let's see how that looks using Azure Event Hubs.
```

Azure Functions come with a powerful system of bindings. So far, we only saw examples of using event sources as trigger bindings. However, Azure also supports a powerful set of so called "input" and "output" bindings. 

### 9. Output Bindings

[ [Runnable Example](https://github.com/pulumi/pulumi-azure/blob/master/examples/queue) ]

Output bindings enable developers to easily forward the data from an Azure Function to an arbitrary destination in a declarative manner. For instance, if a queue handler needs to send a message to another queue, we don't have to use cloud SDKs. Instead, we can return the message-to-be-sent from the callback and wire it to the output queue. Here is a quick example:

``` ts
const incoming = new azure.storage.Queue("queue-in", {
  storageAccountName: storageAccount.name,
});

const outgoing = new azure.storage.Queue("queue-out", {
   storageAccountName: storageAccount.name,
});

incoming.onEvent("NewMessage",  {
   outputs: [outgoing.output("queueOut")],
   callback: async (context, person) => {
       return {
           queueOut: `${person.name} logged into the system`,
       };
   },
});
```

Two elements play together here:

> - The `outputs` property defines an output binding with the name `queueOut` and the destination to `outgoing`.

Reviewer: ```suggestion
- The `outputs` property defines an output binding with the name `queueOut` and the destination to `outgoing`
```

> - The `queueOut` property of the resulting object contains the output message.

Reviewer: ```suggestion
- The `queueOut` property of the resulting object contains the output message
```

> Note that the binding name has to match the output property.

Reviewer: ```suggestion
Note that the binding name must match the output property.
```

### 10. Input Bindings

[ [Runnable Example](https://github.com/pulumi/pulumi-azure/tree/master/examples/table) ]

> Input bindings pull extra bits of information and pass them over as input parameters to the callback. The exact usage depends on the trigger and binding types and might be tricky to get right with JSON configuration files. Here is one example of wiring done in a Pulumi program:

Reviewer: ```suggestion
Input bindings pull extra bits of information and pass them  as input parameters to the callback. The exact usage depends on the trigger and binding types and might be tricky to get right with JSON configuration files. Here is one example of wiring done in a Pulumi program:
```
